presto-testng-services/src/main/java/io/prestosql/testng/services/FlakyAnnotationVerifier.java

```diff
@@ -0,0 +1,124 @@
+                    FlakyAnnotationVerifier.class,
+                    "Failed to process %s: \n%s",
+                    testClass,
+                    getStackTraceAsString(e));
+        }
+    }
+
+    private void reportMethodsWithFlakyAndNoTestAnnotation(ITestClass testClass)
+    {
+        Class<?> realClass = testClass.getRealClass();
+
+        if (realClass.getSuperclass() != null &&
+                "io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy".equals(realClass.getSuperclass().getName())) {
+            // Ignore tempto generated convention tests.
+            return;
+        }
+
+        if (realClass.getName().startsWith("io.prestosql.testng.services.TestFlakyAnnotationVerifier")) {
+            // ignore test of FlakyAnnotationVerifier and internal classes
+            return;
+        }
+
+        List<Method> unannotatedTestMethods = findMethodsWithFlakyAndNoTestAnnotation(realClass);
+        if (!unannotatedTestMethods.isEmpty()) {
+            reportListenerFailure(
+                    FlakyAnnotationVerifier.class,
+                    "Test class %s has methods which are marked as @Flaky but are not explicitly annotated with @Test:%s",
+                    realClass.getName(),
+                    unannotatedTestMethods.stream()
+                            .map(Method::toString)
+                            .collect(joining("\n\t\t", "\n\t\t", "")));
+        }
+    }
+
+    @VisibleForTesting
+    static List<Method> findMethodsWithFlakyAndNoTestAnnotation(Class<?> realClass)
+    {
+        return Arrays.stream(realClass.getMethods())
+                .filter(method -> hasOrInheritsAnnotation(method, Flaky.class))
+                .filter(method -> !hasAnnotation(method, Test.class))
+                .filter(method -> hasOrInheritsAnnotation(method, Test.class))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public void onAfterClass(ITestClass testClass) {}
+
+    private static boolean hasAnnotation(Method method, Class<? extends Annotation> annotationClass)
+    {
+        return method.getAnnotation(annotationClass) != null;
+    }
+
+    private static boolean hasOrInheritsAnnotation(Method method, Class<? extends Annotation> annotationClass)
+    {
+        Optional<Method> currentMethod = Optional.of(method);
+        while (currentMethod.isPresent()) {
+            if (hasAnnotation(currentMethod.get(), annotationClass)) {
+                return true;
+            }
+            currentMethod = getSuperMethod(currentMethod.get());
+        }
+        return false;
+    }
+
+    private static Optional<Method> getSuperMethod(Method method)
+    {
+        Class<?> declaringClass = method.getDeclaringClass().getSuperclass();
+        while (declaringClass != null) {
+            try {
+                return Optional.of(declaringClass.getDeclaredMethod(method.getName(), method.getParameterTypes()));
```

Comment:
If we limit our imagination to public methods only (which is already the case, as you start the search from `realClass.getMethods()`), then you don't need `while + getDeclaredMethod` here.
Calling `superclass.getMethod` will be equally effective.

Note however, that this whole approach deserves as disclaimer like 
https://github.com/prestosql/presto/blob/24153e75240ceda178951a80f9670f431e733292/presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java#L125
-- this is not correct in presence of generics and bridge methods.